lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1593 @@
+          if (auto Member = CMI->getMember()) {
+            if (auto AD = dyn_cast<AccessorDecl>(Member.getDecl())) {
+              if (AD->getAccessorKind() == AccessorKind::Get) {
+                GetterInst = CMI;
+                return true;
+              }
+            }
+          }
+        }
+        
+        ++i;
+      }
+    }
+    return false;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+//                         MARK: Traverser Interface
+//===----------------------------------------------------------------------===//
+
+class VarUsageInfo {
+  
+public:
+
+  SILInstruction *EntryInst;
+  llvm::SetVector<SILInstruction *> *SeenInstructions;
+  
+  SILInstruction *MisusedGetter = nullptr;
+  bool DidRead = false;
+  bool DidModify = false;
+  bool DidDeferInit = false;
+  bool IsWeakCapture = false;
+  VarDecl *InitialReference = nullptr;
+  
+  VarUsageInfo(SILInstruction *EntryInst,
+               llvm::SetVector<SILInstruction *> *SeenInstructions)
+  : EntryInst(EntryInst), SeenInstructions(SeenInstructions) {
+    collectUsageInfo();
+  }
+  
+  void addEntryInst(SILInstruction *I) {
+    EntryInst = I;
+    collectUsageInfo();
+  }
+  
+private:
+  
+  void searchSetterParam() {
+    auto VRT = VariableReadTraverser(EntryInst);
+    if (!VRT.didRead()) {
+      if (auto Fn = EntryInst->getFunction()) {
+        auto SRGT = SetterReadsGetterTraverser(Fn);
+        if (SRGT.findGetterAccess()) {
+          MisusedGetter = SRGT.GetterInst;
+        }
+      }
+    }
+    DidRead = true; // Set this to mute other diagnostics
+  }
+  
+  /// Get the first \c DebugValueInst to reference the value if it is not
+  /// the entry instruction
+  VarDecl *getInitialReference() {
+    auto DVI = dyn_cast<DebugValueInst>(EntryInst);
+    if (!DVI) {
+      return nullptr;
+    }
+    
+    if (auto DebugVD = getVarDecl(EntryInst)) {
+      if (DebugVD->isCaptureList()) {
+        return nullptr;
+      }
+      if (auto ParentStmt = DebugVD->getRecursiveParentPatternStmt()) {
+        if (isa<CaseStmt>(ParentStmt)) {
+          return nullptr;
+        }
+      }
+    }
+    
+    auto ReferencedValue = DVI->getAllOperands()[0].get();
+    
+    DebugValueInst *InitialRefInst = DVI;
+    for (auto Use : ReferencedValue->getUses()) {
+      auto UserDVI = dyn_cast<DebugValueInst>(Use->getUser()); 
+      if (!DVI) {
+        continue;
+      }
+      if (SeenInstructions->count(UserDVI)) {
+        InitialRefInst = UserDVI;
+      }
+    }
+    
+    if (InitialRefInst == DVI) {
+      return nullptr;
+    }
+    
+    return getVarDecl(InitialRefInst);
+  }
+  
+  void collectUsageInfo() {
+    
+    if (DidRead && DidModify) {
+      // Already used; no need for further diagnostic information
+      return;
+    }
+    
+    if (isSetterParam(EntryInst)) {
+      searchSetterParam();
+      // Quit here; setter parameters do not neet further
```

Comment:
... need ...

Reply:
Thanks for catching all these nitpicks! I appreciate your review.